File: leetcode/20_valid-parentheses/stack.go
type Stack []rune

func (s *Stack) IsEmpty() bool {
	return len(*s) == 0
}

func (s *Stack) Push(r rune) {
	*s = append(*s, r)
}

func (s *Stack) Pop() rune {
	if s.IsEmpty() {
		return 0
	}
	r := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return r
}

func isValid(s string) bool {
	expectStack := Stack{}
	for _, r := range s {
		if r == '(' {
			expectStack.Push(')')
		} else if r == '[' {
			expectStack.Push(']')
		} else if r == '{' {
			expectStack.Push('}')
		} else {
			if r != expectStack.Pop() {
				return false
			}
		}
	}
	return expectStack.IsEmpty()
}

File: leetcode/20_valid-parentheses/stack.cpp
// https://leetcode.com/problems/valid-parentheses/submissions/986517956/
class Solution {
 public:
  bool isValid(string s) {
    stack<char> expects;
    for (int i = 0; i < s.size(); i++) {
      if (s[i] == '(') {
        expects.push(')');
      } else if (s[i] == '[') {
        expects.push(']');
      } else if (s[i] == '{') {
        expects.push('}');
      } else if (!expects.empty() && s[i] == expects.top()) {
        expects.pop();
      } else {
        return false;
      }
    }
    return expects.empty();
  }
};
